common/src/main/java/earth/terrarium/heracles/api/tasks/defaults/KillEntityQuestTask.java

```
@@ -55,14 +62,23 @@ public ResourceLocation id() {
         }
 
         @Override
-        public Codec<KillEntityQuestTask> codec(String id) {
-            return RecordCodecBuilder.create(instance -> instance.group(
+        public MapCodec<KillEntityQuestTask> codec(String id) {
+            return RecordCodecBuilder.mapCodec(instance -> instance.group(
                 RecordCodecBuilder.point(id),
-                Codec.STRING.optionalFieldOf("title", "").forGetter(KillEntityQuestTask::title),
-                QuestIcons.CODEC.optionalFieldOf("icon", ItemQuestIcon.AIR).forGetter(KillEntityQuestTask::icon),
-                RestrictedEntityPredicate.CODEC.fieldOf("entity").forGetter(KillEntityQuestTask::entity),
+                Codec.STRING.lenientOptionalFieldOf("title", "").forGetter(KillEntityQuestTask::title),
+                QuestIcons.CODEC.lenientOptionalFieldOf("icon", ItemQuestIcon.AIR).forGetter(KillEntityQuestTask::icon),
+                CODEC.fieldOf("entity").forGetter(KillEntityQuestTask::entity),
                 Codec.INT.fieldOf("amount").orElse(1).forGetter(KillEntityQuestTask::target)
             ).apply(instance, KillEntityQuestTask::new));
         }
     }
+
+    public static final Codec<RestrictedEntityPredicate> CODEC = RecordCodecBuilder.create(instance -> instance.group(
```

This looks like a 1:1 copy of the RestrictedEntityPredicate.CODEC in resourceful lib I think it could use that

---

resourceful lib version uses optionalFieldOf() instead of lenientOptionalFieldOf() which was causing crashes during packet decoding. This was an issue I encountered in various areas throughout the mod.

---

what were the errors that occurred when sending it to the client? because now its just gonna silently fail which isnt the best situation

---

[14:54:41] [Render thread/WARN] [minecraft/ClientCommonPacketListenerImpl]: Client disconnected with reason: Internal Exception: io.netty.handler.codec.DecoderException: Failed to decode packet 'serverbound/minecraft:custom_payload'

It is something to do with not having any default values to be read from "old" or something as far as I can tell in the create() in [KillEntityTaskSettings](https://github.com/JayemCeekay/Heracles/blob/1.21.1/common/src/main/java/earth/terrarium/heracles/api/client/settings/tasks/KillEntityTaskSettings.java).

---

Is that all that it logs? should have some more details, my assumption is its throwing an error on one of the codec entries accessing the registry via registryops

---

It seems to be an issue with using YabnOps.COMPRESSED. Switching to encoding and decoding using YabnOps.INSTANCE lets it decode without issues using the original RestrictedEntityPredicate.CODEC.

---

The difference between COMPRESSED and INSTANCE is that it checks if the codec is keyable and if it is then it doesnt need to include object keys so on the client it can receieve the keys in the constant order so maybe there is a custom codec somewhere that is implementing keyable incorrectly since keys cant be duplicated

---

It might be ComponentSerialization.CODEC in QuestDisplay. Changing it to ComponentSerialization.FLAT_CODEC lets everything load fine with YabnOps.COMPRESSED. However, it does change the way the title and subtitle sections are formatted. For example "title": "{\\"translate\\"\:\"abcd\\"}" vs the older "title": {"translate":"abcd"}

---

That's okay since itemstacks changed between 1.20.x and 1.21.x so the format is already changed